pkg/azureutils/azure_disk_utils.go

```diff
@@ -240,18 +240,39 @@ func GetNodeMaxDiskCount(ctx context.Context, cachedReader client.Reader, nodeNa
 	return GetNodeMaxDiskCountWithLabels(nodeObj.Labels)
 }
 
-func GetNodeRemainingDiskCount(ctx context.Context, cachedReader client.Reader, nodeName string) (int, error) {
-	nodeObj := &v1.Node{}
-	if err := cachedReader.Get(ctx, types.NamespacedName{Name: nodeName}, nodeObj); err != nil {
+// GetNodeRemainingDiskCountApprox returns an approximate capacity to the node
+func GetNodeRemainingDiskCountApprox(ctx context.Context, cachedReader client.Reader, nodeName string) (int, error) {
+	attachments, err := GetAzVolumeAttachmentsForNode(ctx, cachedReader, nodeName, AllRoles)
+	if err != nil {
 		return -1, err
 	}
 
-	// get all replica azvolumeattachments on the node
+	return calculateNodeCapacity(ctx, cachedReader, nodeName, len(attachments))
+}
+
+// GetNodeRemainingDiskCountActual returns the actual capacity to the node
+func GetNodeRemainingDiskCountActual(ctx context.Context, cachedReader client.Reader, nodeName string) (int, error) {
 	attachments, err := GetAzVolumeAttachmentsForNode(ctx, cachedReader, nodeName, AllRoles)
 	if err != nil {
 		return -1, err
 	}
 
+	numberOfAttachedAtts := 0
+	for _, attachment := range attachments {
+		if attachment.Status.State == azdiskv1beta2.Attached {
```

Comment:
We should include disks in the `Attaching` state in the disk count.

Reply:
Currently, we only decrement the capacity when attach succeeded, so the `availableAttachmentsMap` stores actual attachments.
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/f743329bd8249a3bcd1b1685a09820b1b8d5cf0a/pkg/controller/attach_detach.go#L298
 I think including disks in `Attaching` state doesn't align to this logic.

Reply:
Success on attach will decrement the count, so we don't need to decrement for pending states here otherwise we double count. https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/944b31cccffa459b5a6707783a14f26ddc40850a/pkg/controller/attach_detach.go#L298